probe_src/probe_py/analysis.py

```diff
@@ -47,71 +49,77 @@ def first(pid: int, exid: int, tid: int) -> Op:
-
     # Hook up forks/joins
     for node in list(nodes):
-        pid, exid, tid, op = node
-        if False:
-            pass
-        elif isinstance(op, CloneOp):
+        pid, exid, tid, op, op_index = node
+        if isinstance(op, CloneOp):
             if op.flags & CLONE_THREAD:
                 # Spawning a thread links to the current PID and exec epoch
                 target = (pid, exid, op.child_thread_id)
             else:
-                # New process always links to exec epoch 0 and main thread
-                # THe TID of the main thread is the same as the PID
-                target = (op.child_process_id, 0, op.child_process_id)
+                target = (op.child_process_id, 0, 0)
             exec_edges.append((node, first(*target)))
         elif isinstance(op, WaitOp) and op.ferrno == 0 and op.ret > 0:
             # Always wait for main thread of the last exec epoch
             if op.ferrno == 0:
-                target = (op.ret, last_exec_epoch.get(op.ret, 0), op.ret)
+                target = (op.ret, -1, last_exec_epoch.get(op.ret, 0), op.ret)
                 fork_join_edges.append((last(*target), node))
         elif isinstance(op, ExecOp):
             # Exec brings same pid, incremented exid, and main thread
             target = pid, exid + 1, 0
             fork_join_edges.append((node, first(*target)))
-
-    # # Make the main thread exit at the same time as each thread
-    for pid, process in process_tree_prov_log.processes.items():
+            
+    # Make the main thread exit at the same time as each thread
+    for (pid), process in process_tree_prov_log.processes.items():
         for exec_epoch_no, exec_epoch in process.exec_epochs.items():
             for tid, thread in exec_epoch.threads.items():
                 if tid != 0:
-                    fork_join_edges.append((last(pid, exec_epoch_no, tid), last(pid, exec_epoch_no, pid)))
+                    fork_join_edges.append((last(pid, exec_epoch_no, tid), last(pid, exec_epoch_no, 0)))
+
+    process_graph = nx.DiGraph()
+    for node in nodes:
+        process_graph.add_node((node[0], node[1], node[2], node[4]))
+
+    def add_edges(graph, edges):
+        for node0, node1 in edges:
+            node_0 = (node0[0], node0[1], node0[2], node0[4])
+            node_1 = (node1[0], node1[1], node1[2], node1[4])
+            graph.add_edge(node_0, node_1)
+    
+    add_edges(process_graph, program_order_edges)
+    add_edges(process_graph, exec_edges)
+    add_edges(process_graph, fork_join_edges)
+    return process_graph
+
+def digraph_to_pydot_string(process_graph: nx.digraph) -> str:
+    def add_color_to_edges(graph, edges, color):
+        for node0, node1 in edges:
+            node_0 = (node0[0], node0[1], node0[2], node0[4])
+            node_1 = (node1[0], node1[1], node1[2], node1[4])
+            graph[node_0][node_1]['color'] = color
+    
+    add_color_to_edges(process_graph, program_order_edges, 'green')
```

Comment:
Does this work? I don't see how `program_order_edges` gets to this function...

We could label the edges with a flag indicating their type in `provlog_to_digraph` (as in `process_graph.add_edge(node0, node1, label=...)`. However, I would prefer the flag to be an [`IntEnum`](https://docs.python.org/3/library/enum.html#enum.IntEnum) rather than a string, because `IntEnum` specifies a fixed number of options for the value, whereas a string can be anything.

Reply:
Yes it works as I declared the list of edges as global variables. Will look into the labeling of the edges.

Reply:
I declared the list of edges as global variables to be used by both functions. However, I opted to use IntEnum labels for adding colors to the edges, which may render both color and label attributes redundant. Using color appears preferable if the graph is to be visualized using matplotlib in the future. I have pushed the changes for your review. Please let me know if the redundancy needs to be addressed. @charmoniumQ